### PR TITLE
fix #16356

### LINKED ISSUE #16356: Cannot upload a file to a SMB storage with encryption enabled
### Steps to reproduce
1. Install and enable external storage and encryption
2. Create a new SMB mount in the external storage, with a correct configuration available for all users
3. Upload a file inside the SMB mount (bigger than 1MB)

### Expected behaviour
The file is uploaded correctly in the external storage

### Actual behaviour
Only a part of the file is uploaded. It's being cut at 16kB. Smaller files can be uploaded correctly.

screenshot from 2015-05-14 15 27 12

### Server configuration
**Operating system**: ubuntu 14.04

**Web server:** apache 2.4.7

**Database:** mysql

**PHP version:** 5.5.9

**ownCloud version:** OC 8.1 beta 1

**Updated from an older ownCloud or fresh install:** fresh install

**List of activated apps:**
 - encryption: 1.0.0
 - files: 1.1.9
 - files_external: 0.2.3
 - files_locking: 
 - files_sharing: 0.6.1
 - files_trashbin: 0.6.2
 - files_versions: 1.0.5

#### ownCloud log (data/owncloud.log)
```
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u16\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u17\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u18\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u19\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u2\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u20\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u3\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u4\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u5\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u6\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u7\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u8\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}
{"reqId":"dloJ1jRrYoP1sJmvECYh","remoteAddr":"10.40.30.21","app":"encryption","message":"no public key found for user \"u9\", user will not be able to read the file","level":2,"time":"2015-05-14T13:13:49+00:00"}

{"reqId":"JCsAkBwa3TtV3patra2+","remoteAddr":"10.40.30.21","app":"PHP","message":"imagecreatefromjpeg(): gd-jpeg, libjpeg: recoverable error: \b\n at \/home\/jvillafanez\/src\/owncloud\/owncloud2\/lib\/private\/image.php#512","level":3,"time":"2015-05-14T13:27:01+00:00"}

```

### Notes
I've also checked with the storage only available for the current user in order to prevent the "key not found" logs. Those logs doesn't appear, but the behaviour is the same

### PR DESCRIPTION
**Pull request by jknockaert**
Updating `$this->unencryptedSize` should happen after writing each (partial) block such that the tests in `readCache()` work properly and avoid reading overhead.

### CODE REVIEW

**Comment by PVince81** [comment]
Thanks. Would it be possible to write a unit test to reproduce the original issue ?

One idea I had (if possible) would be to have a custom stream wrapper that throws an exception when stream_read is called, to simulate the SMB case.

**Comment by PVince81** [comment]
Fix for #16356

**Comment by jknockaert** [comment]
@PVince81 I have been thinking of a unit test but couldn't come up with something simple. Writing a dummy stream wrapper doesn't sound like a good idea; it would imply that all future updates to the encryption wrapper need to be ported to the dummy wrapper as well. We will introduce far more bugs than we will properly test for.

**Comment by PVince81** [comment]
The dummy wrapper could just be a passthrough to a regular file and only overwrites stream_read.
I think it is possible to use Icewind\Streams\Wrapper for that, it provides a default passthrough wrapper.

**Comment by jknockaert** [comment]
I understand the approach conceptually, but I'm afraid the practical implementation is too far fetched for me.

**Comment by PVince81** [comment]
That's fine, I can give it a try later. Or find an alternative.

**Comment by schiesbn** [comment]
I just tried this PR and while I can't re-produce the original issue with the samba storage I can check if we no longer try to read the file if we write a new file from scratch. But I still observe the behavior that position and unenecryptedSize are out of sync while writing a 6MB large image. I have the feeling that it doesn't happen that often as before, sometimes both have the same value but sometimes they still differ which triggers a read operation.

**Comment by jknockaert** [comment]
@schiesbn Can't think of something else right now.

**Comment by schiesbn** [comment]
@jknockaert I just discovered that you branch wasn't up-to-data on my system after checkout. Tested it again after a git pull on your branch. Looks good now, unencryptedSize and position is always in sync. Awesome! :+1:

**Comment by jknockaert** [comment]
OK. I was already fearing obscure race conditions. Not so--luckily.

**Comment by jknockaert** [comment]
@PVince81 Pls consider upgrading this PR status to review.

**Comment by PVince81** [comment]
The process requires every PR to come with proper unit tests.
We had too many cases in the past where unit tests were missing/forgotten and got loads of regressions afterwards.

**Comment by schiesbn** [comment]
@jknockaert If you want I could take care about the unit test, I already have a idea how this could be done.

**Comment by jknockaert** [comment]
@schiesbn Yes pls go ahead.

**Comment by schiesbn** [comment]
unit test added/updated

**Comment by jknockaert** [comment]
@schiesbn Thanks.

File: tests/lib/files/stream/encryption.php [hunk 1]
```diff
@@ -221,10 +224,15 @@ function dataFilesProvider() {
...
 	public function testWriteReadBigFile($testFile) {
+
 		$expectedData = file_get_contents(\OC::$SERVERROOT . '/tests/data/' . $testFile);
 		// write it
 		$fileName = tempnam("/tmp", "FOO");
 		$stream = $this->getStream($fileName, 'w+', 0);
+		// while writing the file from the beginning to the end we should never try
+		// to read parts of the file. This should only happen if write in the
+		// middle of a file
+		$this->encryptionModule->expects($this->never())->method('decrypt');
```
Comment (PVince81) [inline tests/lib/files/stream/encryption.php]: clever :wink:

**Comment by PVince81** [comment]
Works :+1: 

Now only need a second thumbs up for the unit test @jknockaert :wink:

**Comment by jknockaert** [comment]
Looks good :+1:

**Pull request merged**